File: src/main/kotlin/thc/runtime/CoreVectorMemory.kt
@file:Suppress("UNCHECKED_CAST")
package thc.runtime

/** These are local intrinsics, not vector function or aggregate ABIs. */
internal enum class VectorByteArrayOp(val primitive: String, val scalarOffset: Boolean) {
    INDEX("indexInt32X4Array#", false),
    INDEX_SCALAR("indexInt32ArrayAsInt32X4#", true),
    READ("readInt32X4Array#", false),
    READ_SCALAR("readInt32ArrayAsInt32X4#", true),
    WRITE("writeInt32X4Array#", false),
    WRITE_SCALAR("writeInt32ArrayAsInt32X4#", true);

    val isRead: Boolean get() = this == READ || this == READ_SCALAR
    val isWrite: Boolean get() = this == WRITE || this == WRITE_SCALAR
    val isIndex: Boolean get() = this == INDEX || this == INDEX_SCALAR
    internal fun validateArguments(actual: List<CoreRepresentation>, flags: List<*>) {
        val expected = listOf(CoreVectorMemory.arrayProof, CoreVectorMemory.indexProof) + when {
            isRead -> listOf(CoreVectorMemory.stateProof)
            isWrite -> listOf(CoreVectors.proof32, CoreVectorMemory.stateProof)
            else -> emptyList()
        }
        if (actual.size != expected.size || flags != List(expected.size) { false } ||
            actual.indices.any { !CoreVectorMemory.exact(expected[it], actual[it]) })
            throw RuntimeFault("Vector ByteArray primitive argument representation mismatch: $primitive")
    }
    fun validate(actual: List<CoreRepresentation>, flags: List<*>, result: CoreRepresentation) {
        if (isRead) throw UnsupportedCore("Vector ByteArray read requires an immediate exact case")
        validateArguments(actual, flags)
        if (!CoreVectorMemory.exact(if (isWrite) CoreVectorMemory.stateProof else CoreVectors.proof32, result))
            throw RuntimeFault("Vector ByteArray primitive result representation mismatch: $primitive")
    }
    companion object {
        fun named(name: String): VectorByteArrayOp? = entries.firstOrNull { it.primitive == name }
    }
}

internal data class VectorReadCase(val operation: VectorByteArrayOp, val arguments: List<List<Any?>>,
    val stateBinder: String, val vectorBinder: String, val body: List<Any?>)

internal object CoreVectorMemory {
    val stateProof = CoreRepresentation(CoreKind.VOID, true, true, emptyList())
    internal val arrayProof = CoreRepresentation(CoreKind.OBJECT, true, true, listOf("BoxedRep (Just Unlifted)"))
    internal val indexProof = CoreRepresentation(CoreKind.LONG, true, true, listOf("IntRep"))
    internal fun exact(expected: CoreRepresentation, actual: CoreRepresentation): Boolean =
        actual.present && !actual.isTuple && actual.kind == expected.kind &&
            actual.primReps == expected.primReps && actual.vector == expected.vector

    private fun requireProof(condition: Boolean, detail: String) {
        if (!condition) throw RuntimeFault("Invalid local vector read case: $detail")
    }
    private fun vectorAnnotation(raw: Any?): Boolean {
        val value = raw as? Map<*, *> ?: return false
        val lanes = value["lanes"] as? Number ?: return false
        return lanes.toDouble() == 4.0 && value["element"] == "Int32ElemRep"
    }
    /** The pinned exporter annotates this aggregate with its sole physical VecRep.
     * Validate the original map without admitting it to generic CoreVector.parse. */
    private fun readResult(raw: Any?, binder: Boolean) {
        val value = raw as? Map<*, *> ?: throw RuntimeFault("Missing local vector read result proof")
        val components = value["components"] as? List<*>
        requireProof(value["kind"] == "unknown" && value["aggregate"] == "unboxed-tuple" &&
            value["primReps"] == CoreVectors.proof32.primReps && vectorAnnotation(value["vector"]) &&
            value["evaluated"] is Boolean && (!binder || value["evaluated"] == true) && components?.size == 2,
            "expected exact (# State#, Int32X4# #) proof")
        requireProof(exact(stateProof, CoreRepresentations.parse(components!![0])) &&
            exact(CoreVectors.proof32, CoreRepresentations.parse(components[1])), "result components")
    }
    private fun termUses(value: Any?, id: String): Boolean = when (value) {
        is List<*> -> value.firstOrNull() == "var" && value.getOrNull(1) == id || value.any { termUses(it, id) }
        is Map<*, *> -> value.values.any { termUses(it, id) }
        else -> false
    }
    fun readCase(expr: List<Any?>, constructors: Map<String, Map<String, Any?>>): VectorReadCase? {
        if (expr.firstOrNull() != "case") return null
        val app = expr.getOrNull(1) as? List<*> ?: return null
        if (app.firstOrNull() != "app") return null
        val function = app.getOrNull(1) as? List<*> ?: return null
        if (function.firstOrNull() != "prim") return null
        val operation = (function.getOrNull(1) as? String)?.let(VectorByteArrayOp::named) ?: return null
        if (!operation.isRead) return null
        val arguments = (app.getOrNull(2) as? List<*>)?.map {
            it as? List<Any?> ?: throw RuntimeFault("Invalid local vector read argument")
        } ?: throw RuntimeFault("Missing local vector read arguments")
        val flags = app.getOrNull(3) as? List<*> ?: throw RuntimeFault("Missing local vector read flags")
        operation.validateArguments(arguments.map(CoreRepresentations::expression), flags)
        readResult((app.getOrNull(6) as? Map<*, *>)?.get("rep"), false)
        val whole = expr.getOrNull(2) as? String ?: throw RuntimeFault("Missing local vector read case binder")
        val metadata = expr.getOrNull(4) as? Map<*, *> ?: throw RuntimeFault("Missing local vector read metadata")
        val binder = metadata["binder"] as? Map<*, *> ?: throw RuntimeFault("Missing local vector read binder metadata")
        requireProof(binder["id"] == whole && binder["lifted"] == false && "joinValueArity" !in binder,
            "whole-tuple binder identity/levity")
        readResult(binder["rep"], true)
        val alternatives = expr.getOrNull(3) as? List<*> ?: throw RuntimeFault("Missing local vector read alternatives")
        requireProof(alternatives.size == 1, "requires one tuple alternative")
        val alternative = alternatives.single() as? List<*> ?: throw RuntimeFault("Invalid local vector read alternative")
        val constructor = (alternative.getOrNull(1) as? String)?.let(constructors::get)
        requireProof(alternative.firstOrNull() == "data" && constructor?.get("kind") == "unboxed-tuple" &&
            (constructor?.get("arity") as? Number)?.toDouble() == 2.0, "requires a registered tuple2 constructor")
        val ids = alternative.getOrNull(2) as? List<*> ?: throw RuntimeFault("Missing local vector read pattern ids")
        requireProof(ids.size == 2 && ids.all { it is String } && ids.distinct().size == 2 && whole !in ids,
            "pattern binder identities")
        val records = (alternative.getOrNull(4) as? Map<*, *>)?.get("binders") as? List<*>
        requireProof(records?.size == 2, "missing ordered pattern metadata")
        listOf(stateProof, CoreVectors.proof32).forEachIndexed { i, expected ->
            val record = records!![i] as? Map<*, *> ?: throw RuntimeFault("Invalid local vector read pattern metadata")
            requireProof(record["id"] == ids[i] && record["lifted"] == false && "joinValueArity" !in record &&
                exact(expected, CoreRepresentations.parse(record["rep"])) &&
                (record["rep"] as? Map<*, *>)?.get("evaluated") == true, "pattern binder representation")
        }
        val body = alternative.getOrNull(3) as? List<Any?> ?: throw RuntimeFault("Missing local vector read continuation")
        requireProof(body.isNotEmpty(), "empty continuation")
        requireProof(!termUses(body, whole), "whole tuple binder escapes")
        return VectorReadCase(operation, arguments, ids[0] as String, ids[1] as String, body)
    }
}

File: scripts/core_vector_memory.py
"""Exact local Int32X4/Word32X4/FloatX4 ByteArray intrinsics; no vector aggregate transport.

The pinned exporter places its sole physical vector annotation on the logical
State/vector tuple too. Only an immediate, exactly checked read case accepts it.
"""
from core_vectors import VECTOR32_REP, VECTOR_WORD32_REP, VECTOR_FLOAT_REP, signature_matches

STATE = dict(kind='void', primReps=[], evaluated=True)
ARRAY = dict(kind='object', primReps=['BoxedRep (Just Unlifted)'], evaluated=True)
INDEX = dict(kind='long', primReps=['IntRep'], evaluated=True)
SIGNED_READS = {'readInt32X4Array#', 'readInt32ArrayAsInt32X4#'}
SIGNED_INDICES = {'indexInt32X4Array#', 'indexInt32ArrayAsInt32X4#'}
SIGNED_WRITES = {'writeInt32X4Array#', 'writeInt32ArrayAsInt32X4#'}
SIGNED_OPERATIONS = SIGNED_READS | SIGNED_INDICES | SIGNED_WRITES
UNSIGNED_READS = {'readWord32X4Array#', 'readWord32ArrayAsWord32X4#'}
UNSIGNED_INDICES = {'indexWord32X4Array#', 'indexWord32ArrayAsWord32X4#'}
UNSIGNED_WRITES = {'writeWord32X4Array#', 'writeWord32ArrayAsWord32X4#'}
UNSIGNED_OPERATIONS = UNSIGNED_READS | UNSIGNED_INDICES | UNSIGNED_WRITES
FLOAT_READS = {'readFloatX4Array#', 'readFloatArrayAsFloatX4#'}
FLOAT_INDICES = {'indexFloatX4Array#', 'indexFloatArrayAsFloatX4#'}
FLOAT_WRITES = {'writeFloatX4Array#', 'writeFloatArrayAsFloatX4#'}
FLOAT_OPERATIONS = FLOAT_READS | FLOAT_INDICES | FLOAT_WRITES
READS = SIGNED_READS | UNSIGNED_READS | FLOAT_READS
INDICES = SIGNED_INDICES | UNSIGNED_INDICES | FLOAT_INDICES
WRITES = SIGNED_WRITES | UNSIGNED_WRITES | FLOAT_WRITES
OPERATIONS = READS | INDICES | WRITES
VECTOR_PROOFS = {**dict.fromkeys(SIGNED_OPERATIONS, VECTOR32_REP),
                 **dict.fromkeys(UNSIGNED_OPERATIONS, VECTOR_WORD32_REP),
                 **dict.fromkeys(FLOAT_OPERATIONS, VECTOR_FLOAT_REP)}


def require(condition, detail):
    if not condition:
        raise ValueError('Invalid local vector memory intrinsic: ' + detail)


def vector_proof(name):
    require(name in OPERATIONS, 'unknown memory operation')
    return VECTOR_PROOFS[name]


def representation(expr):
    index = {'var': 2, 'lit': 3, 'app': 6, 'lam': 3, 'let': 4,
             'case': 4, 'con': 3, 'prim': 2, 'void': 1}.get(expr[0])
    metadata = expr[index] if index is not None and len(expr) > index else None
    return metadata.get('rep') if isinstance(metadata, dict) else None


def exact(expected, actual):
    return (isinstance(actual, dict) and type(actual.get('evaluated')) is bool and
            signature_matches(expected, actual) and
            (expected.get('kind') != 'vector' or
             type(actual.get('vector', {}).get('lanes')) is int))


def validate_arguments(name, arguments, flags):
    vector = vector_proof(name)
    expected = [ARRAY, INDEX] + ([STATE] if name in READS else [vector, STATE] if name in WRITES else [])
    require(isinstance(arguments, list) and len(arguments) == len(expected), 'argument arity')
    require(isinstance(flags, list) and len(flags) == len(expected) and all(f is False for f in flags), 'unlifted flags')
    for argument, wanted in zip(arguments, expected):
        require(isinstance(argument, list) and argument, 'argument expression')
        require(exact(wanted, representation(argument)), 'argument representation')


def validate_direct(name, arguments, flags, result):
    require(name in INDICES | WRITES, 'read requires an immediate exact case')
    validate_arguments(name, arguments, flags)
    require(exact(STATE if name in WRITES else vector_proof(name), result), 'result representation')


def read_result(proof, vector, binder=False):
    require(isinstance(proof, dict), 'missing read result proof')
    shape = proof.get('vector')
    require(isinstance(shape, dict) and type(shape.get('lanes')) is int and
            shape == vector['vector'], 'aggregate vector annotation')
    require(proof.get('kind') == 'unknown' and proof.get('aggregate') == 'unboxed-tuple' and
            proof.get('primReps') == vector['primReps'] and
            type(proof.get('evaluated')) is bool and (not binder or proof['evaluated']), 'read result shape')
    fields = proof.get('components')
    require(isinstance(fields, list) and len(fields) == 2 and
            exact(STATE, fields[0]) and exact(vector, fields[1]), 'read result components')


def term_uses(value, identity):
    if isinstance(value, list):
        return ((len(value) > 1 and value[0] == 'var' and value[1] == identity) or
                any(term_uses(child, identity) for child in value))
    if isinstance(value, dict):
        return any(term_uses(child, identity) for child in value.values())
    return False


def read_case(expr, constructors):
    """None for other syntax; recognized-invalid read cases always raise."""
    if not isinstance(expr, list) or not expr or expr[0] != 'case':
        return None
    app = expr[1] if len(expr) > 1 else None
    if not isinstance(app, list) or len(app) < 2 or app[0] != 'app':
        return None
    function = app[1]
    if (not isinstance(function, list) or len(function) < 2 or function[0] != 'prim' or
            function[1] not in READS):
        return None
    require(len(app) > 6 and len(expr) > 4, 'missing application/case fields')
    name = function[1]
    validate_arguments(name, app[2], app[3])
    vector = vector_proof(name)
    read_result(representation(app), vector)
    whole = expr[2]
    require(isinstance(whole, str), 'case binder id')
    metadata = expr[4]
    require(isinstance(metadata, dict), 'case metadata')
    binder = metadata.get('binder')
    require(isinstance(binder, dict) and binder.get('id') == whole and binder.get('lifted') is False and binder.get('coercion') is False and
            'joinValueArity' not in binder, 'whole-tuple binder identity/levity')
    read_result(binder.get('rep'), vector, True)
    alternatives = expr[3]
    require(isinstance(alternatives, list) and len(alternatives) == 1, 'requires one tuple alternative')
    alternative = alternatives[0]
    require(isinstance(alternative, list) and len(alternative) > 4, 'alternative metadata')
    require(alternative[0] == 'data' and isinstance(alternative[1], str), 'requires a tuple data alternative')
    constructor = constructors.get(alternative[1])
    require(isinstance(constructor, dict) and constructor.get('kind') == 'unboxed-tuple' and
            type(constructor.get('arity')) is int and constructor['arity'] == 2, 'requires a registered tuple2 constructor')
    ids = alternative[2]
    require(isinstance(ids, list) and len(ids) == 2 and all(isinstance(i, str) for i in ids) and
            len(set(ids)) == 2 and whole not in ids, 'pattern binder identities')
    records = alternative[4].get('binders') if isinstance(alternative[4], dict) else None
    require(isinstance(records, list) and len(records) == 2, 'ordered pattern metadata')
    for identity, wanted, record in zip(ids, [STATE, vector], records):
        require(isinstance(record, dict) and record.get('id') == identity and record.get('lifted') is False and record.get('coercion') is False and
                'joinValueArity' not in record and exact(wanted, record.get('rep')) and
                record['rep']['evaluated'] is True, 'pattern binder representation')
    body = alternative[3]
    require(isinstance(body, list) and body, 'continuation')
    require(not term_uses(body, whole), 'whole tuple binder escapes')
    return dict(operation=name, arguments=app[2], stateBinder=ids[0], vectorBinder=ids[1], body=body)
